core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -364,77 +375,126 @@ private void handleResponseWithoutContent(RetryConfig<HttpResponse> config, Clie
-                                       truncatingHttpResponse.abort();
-                                       handleRetryDecision(decision, ctx, derivedCtx, rootReqDuplicator,
-                                                           originalReq, returnedRes, future, duplicated);
-                                       return null;
-                                   });
-                } catch (Throwable cause) {
-                    duplicator.abort(cause);
-                    handleException(ctx, rootReqDuplicator, future, cause, false);
-                }
-                return;
+    private void handleAggregatedResponse(RetryConfig<HttpResponse> retryConfig, ClientRequestContext ctx,
+                                          HttpRequestDuplicator rootReqDuplicator,
+                                          HttpRequest originalReq, HttpResponse returnedRes,
+                                          CompletableFuture<HttpResponse> future,
+                                          ClientRequestContext derivedCtx,
+                                          AggregatedHttpResponse aggregatedRes) {
+        if (retryConfig.needsContentInRule()) {
+            final RetryRuleWithContent<HttpResponse> ruleWithContent = retryConfig.retryRuleWithContent();
+            assert ruleWithContent != null;
+            try {
+                ruleWithContent.shouldRetry(derivedCtx, aggregatedRes.toHttpResponse(), null)
+                               .handle((decision, cause) -> {
+                                   warnIfExceptionIsRaised(ruleWithContent, cause);
+                                   handleRetryDecision(
+                                           decision, ctx, derivedCtx, rootReqDuplicator, originalReq,
+                                           returnedRes, future, aggregatedRes.toHttpResponse());
+                                   return null;
+                               });
+            } catch (Throwable cause) {
+                handleException(ctx, rootReqDuplicator, future, cause, false);
             }
-            final HttpResponse response0;
-            if (aggregatedRes != null) {
-                response0 = aggregatedRes.toHttpResponse();
+            return;
+        }
+        handleResponseWithoutContent(retryConfig, ctx, rootReqDuplicator, originalReq, returnedRes,
+                                     future, derivedCtx, aggregatedRes.toHttpResponse(), null);
+    }
+
+    private static void completeLogIfIncomplete(AggregatedHttpResponse response, ClientRequestContext ctx) {
+        if (!ctx.log().isAvailable(RequestLogProperty.REQUEST_FIRST_BYTES_TRANSFERRED_TIME)) {
+            final RequestLogBuilder logBuilder = ctx.logBuilder();
+            logBuilder.endRequest();
+            logBuilder.responseHeaders(response.headers());
+            if (!response.trailers().isEmpty()) {
+                logBuilder.responseTrailers(response.trailers());
+            }
+            logBuilder.endResponse();
```

Comment:
Shouldn't we also set the response content to null in case it's deferred?
https://github.com/line/armeria/blob/5d07d39ed516b32b6839bcc803578612efbb7a44/core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java#L542-L548

Reply:
Do we have to call `responseContent(null, null)` and `responseContentPreview(null)` when we end `RequestLog` manually? I'd like to clarify in which cases they should be set to null.

Reply:
The `RequestLog` might not be complete because the properties are deferred when content previewing client is used?
https://github.com/line/armeria/blob/ce03ff4333de56c3facf515bf0587245ca36aa4d/core/src/main/java/com/linecorp/armeria/client/logging/ContentPreviewingClient.java#L160

Reply:
Would it be simpler to fix `ContentPreviewingClient` to set `responseContentPreview(null)` if an exception is raised by the delegate? For example:
```java
try {
    ctx.logBuilder().defer(RequestLogProperty.RESPONSE_CONTENT_PREVIEW);
    final HttpResponse res = unwrap().execute(ctx, req);
    return setUpResponseContentPreviewer(contentPreviewerFactory, ctx, res, responsePreviewSanitizer);
} catch (Throwable e) {
    ctx.logBuilder().responseContentPreview(null);
    return HttpResponse.ofFailure(e);
}
```

It seems other cases are covered by 
https://github.com/line/armeria/blob/a5be69d99f03f892abfeca017597340e2b749eea/core/src/main/java/com/linecorp/armeria/internal/logging/ContentPreviewingUtil.java#L116-L132

Reply:
That probably doesn't work in this situation?
UserClient -> content previewing client -> retrying client -> short circuit client

Reply:
I added `ContentPreviewingClientLogCompletionTest` to check if child `RequestLog`s are complete. According to the test results, all logs are completed. 

If there’s a corner case not covered by that test, please let me know.

Reply:
Thanks for adding the test. 😉